### ISSUE TITLE
How to create Ground Truth and Prediction txt files

### ISSUE THREAD

**Issue by hiteshkhanna31** [rafaelpadilla/Object-Detection-Metrics#104]
I am using Tensor flow 2 for YOLOV3 (https://github.com/pythonlessons/TensorFlow-2.x-YOLOv3) and i want to evaluate it on MS COCO validation dataset. How should the prediction code be modifed so that i have a detection txt generate for every image and  how can i make Groundtruth txt from instance_val2017.json files available in the dataset.

I want to use this Object detection metrics to calculate mAP.

Appreciate if someone can help as i am beginner in Computer vision domain.

**Comment by rafaelpadilla** [comment]
@hiteshkhanna31 

You need to check the format of the groundtruth bounding boxes used in the training set. Then check the bounding box format output by your network.

Follow the instructions and the parameters you should use [here](https://github.com/rafaelpadilla/Object-Detection-Metrics/blob/master/README.md#how-to-use-this-project)

**Comment by renhongtao** [comment]
you can use others  python tool do this job.Down  is my  code  ,copy from others and change a little code

# Script to convert  voc format to   name    left  top right  down format

# Sample format
# <annotation>
#	 <folder>_image_fashion</folder>
#	 <filename>brooke-cagle-39574.jpg</filename>
#	 <size>
#		 <width>1200</width>
#		 <height>800</height>
#		 <depth>3</depth>
#	 </size>
#	 <segmented>0</segmented>
#	 <object>
#		 <name>head</name>
#		 <pose>Unspecified</pose>
#		 <truncated>0</truncated>
#		 <difficult>0</difficult>
#		 <bndbox>
#			 <xmin>549</xmin>    left
#			 <ymin>251</ymin>     top
#			 <xmax>625</xmax>       right 
#			 <ymax>335</ymax>        down
#		 </bndbox>
#	 </object>
# <annotation>
import os
import xml.etree.cElementTree as ET
import argparse, sys
import time
from PIL import Image

class VOCXml(object):
	DEBUG = True
	SHOW_VOC_BOX_VALUE = True
	JPG_DIR="home/rht"
	#write in lowercase
	CLASS_MAPPING = {"diaoche":"crane",
	"crane":"crane",
	"excavator":"excavator",
	"person":"person",
	"bulldozer":"bulldozer", 
	"shigongxie":"4",
	"dizhuan":"5",
	"geological_survey_machine": "geological_survey_machine",
	"directional_rig":"directional_rig"}

	def convert(self,size, box):
		if VOCXml.SHOW_VOC_BOX_VALUE: print("w:%d h:%d voc:%s"%(size[0],size[1],box))
		dw = 1./(size[0])
		dh = 1./(size[1])
		x = (box[0] + box[1])/2.0 - 1  #top 
		y = (box[2] + box[3])/2.0 - 1
		w = box[1] - box[0]
		h = box[3] - box[2]
		x = x*dw
		w = w*dw
		y = y*dh
		h = h*dh
		return (x,y,w,h)
	def ask_if_exit(self):
		ans = input ("if quit [y or n] ?")
		if ans.find("y") > -1:
			exit(0)

	def get_img_weight_height(self,filepath):
		"""
			return 
			weight,height
		"""
		img = Image.open(filepath)
		return img.size
		

	def convert_file(self,xml_label_file):
		in_file = open(xml_label_file)
		file_name = xml_label_file.split(".xml")[0].split("/")[-1]+".txt"
		file_path_name = os.path.join(VOCXml.YOLO_LABEL_DIR, file_name)
		if self.DEBUG:	print ("process:%s"%(xml_label_file)) 

		out_file = open(file_path_name, 'w')
		tree=ET.parse(in_file)
		root = tree.getroot()
		size = root.find('size')
		w = int(size.find('width').text)
		h = int(size.find('height').text)
		while 0==w or 0 == h:
			print("Error: %s is xml data is error! why weight or height is 0?"%(xml_label_file))
							
			while VOCXml.JPG_DIR is None or not os.path.exists(VOCXml.JPG_DIR):			
				VOCXml.JPG_DIR=input("plseas set right img dir:").strip()
			img_file=os.path.join(VOCXml.JPG_DIR,xml_label_file.split("/")[-1][:-4]+".jpg")
			w,h = self.get_img_weight_height(img_file)
			input("img_file {} w:{} h:{} \nwait key to continue...".format(img_file,w,h))
			
			
		for obj in root.iter('object'):
			difficult = obj.find('difficult').text
			cls = obj.find('name').text.lower()
			if VOCXml.CLASS_MAPPING.get(cls) is None:
				print("{} is not in classesMap".format(cls))
				ask_if_exit()
				continue
			cls_id = VOCXml.CLASS_MAPPING.get(cls)
			xmlbox = obj.find('bndbox')
			b = (float(xmlbox.find('xmin').text)
				, float(xmlbox.find('ymin').text)
				, float(xmlbox.find('xmax').text)
				, float(xmlbox.find('ymax').text))
			if not (b[0] < b[2] and b[1] < b[3]):print (b)
			assert (b[0] < b[2] and b[1] < b[3])
			left_right_top_bottom = b #self.convert((w,h), b)
			#out_file.write(str(cls_id) + " " + " ".join([str(a) for a in left_right_top_bottom]) + '\n')
			out_file.write("%s %s\n"%(cls_id," ".join([str(a) for a in left_right_top_bottom])))
		print("Processing complete for file: {}".format(xml_label_file))


	def start(self,voc_src_dir,yolo_dst_dir):
		if not os.path.exists(yolo_dst_dir):
			os.makedirs(yolo_dst_dir)
		for voc_label_filename in os.listdir(voc_src_dir):
			if voc_label_filename.endswith('xml'):
				file = os.path.join(voc_src_dir, voc_label_filename) 
				self.convert_file(file)
			else:
				print("Skipping file: {}".format(voc_label_filename))


if __name__ == "__main__":
	parser=argparse.ArgumentParser()
	parser.add_argument('-voc','--voc_label_dir', help='set dst dir of voc lable xml file')
	parser.add_argument('-dst','--dst_dir', help='set yolo lable files dir,if not set,use voc dir as out put ')
	parser.add_argument('-jpg','--jpg_dir',help="some time the labeImage tool don't get right img size need to read img get w h")
	args=parser.parse_args()
	
	voc = VOCXml()
	VOCXml.YOLO_LABEL_DIR=args.dst_dir
	VOCXml.VOC_LABEL_DIR_DST=args.voc_label_dir
	VOCXml.JPG_DIR = args.jpg_dir
	if not os.path.exists(VOCXml.VOC_LABEL_DIR_DST):
		print("yolo label dir is not exist !")
	if VOCXml.YOLO_LABEL_DIR is None:
		VOCXml.YOLO_LABEL_DIR = VOC_LABEL_DIR_DST
	
	input_str = input("are you sure the index:{}  [y/n]".format(VOCXml.CLASS_MAPPING)).lower() 
	print("you input is",input_str)
	if input_str.find("y") > -1:		
	
		voc.start(VOCXml.VOC_LABEL_DIR_DST,VOCXml.YOLO_LABEL_DIR)

**Comment by rafaelpadilla** [comment]
Dear @hiteshkhanna31,

Has your question been answered? Can I close this issue?

**Comment by hiteshkhanna31** [comment]
Thank you for your support..
